### PR TITLE
fix(webrtc): gracefully handle resets of individual connections

### LINKED ISSUE #3574: WebRTC listener always closes when any browser client disconnects
## Summary

**Issue**: When I start a libp2p server with **WebRTC** transport, then connect with a few browser clients, when any of the clients drop off, the swarm listener is closed instead of remaining open to handle remaining or subsequent browser connections.

**Background**: When I was creating a long lived WebRTC server using `ping` and `keep_alive`, every time I closed the browser, libp2p would also close the WebRTC listener.

[Discussion](https://discuss.libp2p.io/t/does-keep-alive-behaviour-work-with-webrtc-transport/1849/1) - Trying to work through whether this is a bug or intended behaviour...
[Reproducable Example: https://github.com/DougAnderson444/webrtc-ping](https://github.com/DougAnderson444/webrtc-ping) - A simple ping server and web client

## Expected behaviour

I expected the WebRTC server listener to keep listening as browser clients come and go. 

As an aside, I originally expected the `keep_alive` behaviour to ensure the Transport listener kept listening, but after investigation I realize the `keep_alive` behaviour was not preventing the `webrtc::tokio::Tranport` listener from closing in the event of a Mux Error.

## Actual behaviour

When a browser client disappears (refresh, browser closes, doesn't matter) the WebRTC server receives a `UDPMuxEvent::Error` which closes the listener. The Transport closes the listener on any `UDPMuxEvent::Error`, even if we want to keep listening.

<details><summary>Debug Output</summary>
<p>

```
Event: ListenerClosed { 
listener_id: ListenerId(16861393669339383903), addresses: ["/ip6/::1/udp/42069/webrtc/certhash/uEiDHVx-evVihAOvIsWCX0Za1fcwbaHMBZKPbkyeUodTV2A"], 
reason: Err(Custom { kind: Other, error: 
UDPMux(Os { code: 10054, kind: 
ConnectionReset, 
message: "An existing connection was forcibly closed by the remote host." }) }) }

=== Full Debug Output: ===

[2023-03-08T17:21:48Z DEBUG webrtc_sctp::association] [] recving 32 bytes
[2023-03-08T17:21:48Z WARN  webrtc_sctp::association] [] failed to handle_inbound: ErrChunk
[2023-03-08T17:21:48Z ERROR libp2p_webrtc::tokio::udp_mux] Could not read udp packet: An existing connection was forcibly 
closed by the remote host. (os error 10054)
[2023-03-08T17:21:48Z DEBUG libp2p_swarm] Listener ListenerId(16492161634537859209); Closed by Err(Custom { kind: Other, error: UDPMux(Os { code: 10054, kind: ConnectionReset, message: "An existing connection was forcibly closed by the remote host." }) }).
🎇  Event: ListenerClosed { listener_id: ListenerId(16492161634537859209), addresses: ["/ip6/2607:fea8:fec0:7337::d93a/udp/42069/webrtc/certhash/uEiC8o1MohfA51FM3ric-jBs8hqXreVgbxxzULV9ERtFK3g", "/ip6/2607:fea8:fec0:7337:6b21:9df6:d46e:29d3/udp/42069/webrtc/certhash/uEiC8o1MohfA51FM3ric-jBs8hqXreVgbxxzULV9ERtFK3g", "/ip6/2607:fea8:fec0:7337:6860:ab4c:4715:f9d8/udp/42069/webrtc/certhash/uEiC8o1MohfA51FM3ric-jBs8hqXreVgbxxzULV9ERtFK3g", "/ip6/::1/udp/42069/webrtc/certhash/uEiC8o1MohfA51FM3ric-jBs8hqXreVgbxxzULV9ERtFK3g"], reason: Err(Custom { kind: Other, error: UDPMux(Os { code: 10054, kind: ConnectionReset, message: "An existing connection was forcibly closed by the remote host." }) }) }

[2023-03-08T17:21:48Z DEBUG webrtc_sctp::association::association_internal] [] state change: 'Established' => 'Closed'    
[2023-03-08T17:21:48Z DEBUG webrtc_sctp::association::association_internal] [] closing association..
[2023-03-08T17:21:48Z DEBUG webrtc_sctp::association::association_internal] [] association closed
[2023-03-08T17:21:48Z DEBUG webrtc_sctp::association::association_internal] [] stats nDATAs (in) : 9
[2023-03-08T17:21:48Z DEBUG webrtc_sctp::association::association_internal] [] stats nSACKs (in) : 9
[2023-03-08T17:21:48Z DEBUG webrtc_sctp::association::association_internal] [] stats nT3Timeouts : 0
[2023-03-08T17:21:48Z DEBUG webrtc_sctp::association::association_internal] [] stats nAckTimeouts: 2
[2023-03-08T17:21:48Z DEBUG webrtc_sctp::association::association_internal] [] stats nFastRetrans: 0
[2023-03-08T17:21:48Z DEBUG webrtc_sctp::association] [] read_loop exited
[2023-03-08T17:21:48Z INFO  libp2p_webrtc::tokio::substream::drop_listener] Substream 1 dropped without graceful close, sending Reset
🎇  Event: Behaviour(Ping(Event { peer: PeerId("12D3KooWQiqhey7Q7p43KhhiE2vhfG797eGzZore48bFTUpYgrbS"), result: Err(Unsupported) }))

[2023-03-08T17:21:48Z DEBUG webrtc_sctp::association] [] write_loop exited
[2023-03-08T17:21:48Z DEBUG libp2p_webrtc::tokio::connection] a DropListener failed: Stream closed
[2023-03-08T17:21:53Z INFO  webrtc_ice::agent::agent_internal] [controlled]: Setting new connection state: Disconnected
[2023-03-08T17:21:53Z INFO  webrtc::peer_connection] ICE connection state changed: disconnected
[2023-03-08T17:21:53Z INFO  webrtc::peer_connection] peer connection state changed: disconnected


```
</p>
</details>

## Possible Solution

Match on ConnectionReset Error type in `libp2p_webrtc::tokio::transport` in `impl Stream for ListenStream`  at `fn poll_next`:
https://github.com/libp2p/rust-libp2p/blob/babf7e375310b651342e782880aadb65c620c00e/transports/webrtc/src/tokio/transport.rs#L351-L354

Instead of closing the listener, don't close the listener if the connection is reset by a client:

```rs
Poll::Ready(UDPMuxEvent::Error(e)) => {
    match e.kind() {
        // If client drops off and resets the connection, do nothing with the ListenStream
        ErrorKind::ConnectionReset => {
            log::debug!("ConnectionReset, keep ListenStream open {e:?}");
            todo!("Trigger poll to close any connections?")
        }
        // close the listener on any other kind of Mux error 
        _ => {
            self.close(Err(Error::UDPMux(e)));
            continue;
        }
    }
}
```

## Version

- libp2p version (version number, commit, or branch): 1.65.0

## Would you like to work on fixing this bug?

Yes, I have forked the repo, made the change, and the bug disappears upon doing so. But I'm not sure about any undesired side effects of this change and would appreciate design input for a change like this!

### PR DESCRIPTION
**Pull request by DougAnderson444**
## Description

Fixes https://github.com/libp2p/rust-libp2p/issues/3574

## Notes

When Connection is Reset in UDP Mux, do not emit an Error event, so that the listener stays alive.

## Links to any relevant issues

https://github.com/libp2p/rust-libp2p/issues/3574

## Open Questions

When the connection is reset, is there any need to also Close the Connection from the server side? Or will the polling take care of this next time around (it does with ping behaviour)?
 
I didn't write a test for this -- How do you write a test that fires a `std::io::error` for `ErrorKind::ConnectionReset` on the `udp_sock`?

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] ~~I have added tests that prove my fix is effective or that my feature works~~
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: transports/webrtc/CHANGELOG.md
Comment (thomaseizinger) [inline transports/webrtc/CHANGELOG.md]: Can we undo these formatting changes please?

File: transports/webrtc/CHANGELOG.md
Comment (thomaseizinger) [inline transports/webrtc/CHANGELOG.md]:
```suggestion
# 0.4.0-alpha.3 - unreleased

- Gracefully handle errors on individual connections, avoiding shutdown of the entire listener upon disconnect of a single client.
  See [PR 3575].
  
[PR 3575]: https://github.com/libp2p/rust-libp2p/pull/3575
```

File: transports/webrtc/src/tokio/udp_mux.rs
Comment (thomaseizinger) [inline transports/webrtc/src/tokio/udp_mux.rs]: Can we include any connection details here?

File: transports/webrtc/src/tokio/udp_mux.rs
Comment (thomaseizinger) [inline transports/webrtc/src/tokio/udp_mux.rs:401]: I'd prefer if we stick to this pattern matching style, it results in a flatter code hierarchy.

**Review by thomaseizinger — DISMISSED** [review]
Thanks! Some suggestions.

File: transports/webrtc/CHANGELOG.md
Comment (DougAnderson444) [reply transports/webrtc/CHANGELOG.md]: Yes definitely, I didn't see those until after I had already committed it. They should be rolled back now

Comment (thomaseizinger) [reply transports/webrtc/CHANGELOG.md]: Still visible for me.

File: transports/webrtc/src/tokio/udp_mux.rs
Comment (DougAnderson444) [reply transports/webrtc/src/tokio/udp_mux.rs]: I've been looking, but I still don't see how we could correlate this `ConnectionReset` error to any given `connection`.

The operating system error (code 10054) gives us no information about the connection or how to identify which connection (it actually says "**_An_** existing connection was forcibly closed" -- but not which one). 

It appears the Muxer needs a ufrag to identify the connections, which we also don't have in this context. 

So I'm not seeing a way to identify which connection was reset within the muxer

File: transports/webrtc/src/tokio/udp_mux.rs
Comment (DougAnderson444) [reply transports/webrtc/src/tokio/udp_mux.rs:401]: Yes agreed, I'll reformat it

File: transports/webrtc/CHANGELOG.md
Comment (DougAnderson444) [reply transports/webrtc/CHANGELOG.md]: Well said

File: transports/webrtc/src/tokio/udp_mux.rs
Comment (DougAnderson444) [reply transports/webrtc/src/tokio/udp_mux.rs]: To add to this: Eventually the reset connection is polled by the swarm behaviour and then closed with `SwarmEvent::ConnectionClosed`. So information on the connection can be obtained from the Swarm when this happens, if desired.

**Review by mxinden — COMMENTED** [review]
First off, great work @DougAnderson444 for debugging and proposing a patch. This is great help. I am sorry you ran into this. Thank you!

(As an aside, also great to see more usage of the WebRTC implementation in rust-libp2p.)

I am still working myself through the many layers here. Sharing my thoughts thus far early.

---

Our `UDPMuxNewAddr` is a fork of `webrtc-rs`'s `UDPMuxDefault`.

After a quick search, it seems that they don't filter for a `ConnectionReset`, but instead `break` the loop for the socket. I would expect that to result in the same behavior as you reported @DougAnderson444. Has any of you looked further into this? Why is it that they don't need to handle the `ConnectionReset`?

https://github.com/webrtc-rs/webrtc/blob/5859550cac046376bd4c7a252e79199c54f44bb8/ice/src/udp_mux/mod.rs#L205-L209

---

I am a bit thrown off by the fact that we get a _connection_ reset error on UDP socket, i.e. a connectionless protocol. Also, how should the receiving side detect a connection reset error if the other side (browser) just disappears (i.e. is closed) without sending any further data.

Off the top of my head I would explain this in the following way: on a UDP socket, send and receive share the same error slot. A previous send on the UDP socket caused an error, reported as a `ConnectionReset`. Our call to `poll_recv_from` surfaces that error from the previous send.

Any alternative explanations?

**Comment by mxinden** [comment]
Again thinking out loud:

> Our `UDPMuxNewAddr` is a fork of `webrtc-rs`'s `UDPMuxDefault`.
> 
> After a quick search, it seems that they don't filter for a `ConnectionReset`, but instead `break` the loop for the socket. I would expect that to result in the same behavior as you reported @DougAnderson444. Has any of you looked further into this? Why is it that they don't need to handle the `ConnectionReset`?
> 
> https://github.com/webrtc-rs/webrtc/blob/5859550cac046376bd4c7a252e79199c54f44bb8/ice/src/udp_mux/mod.rs#L205-L209

`webrt-rs` might have a different polling behavior where the `ConnectionReset` is surfaced through the `send_to` instead of the `recv_from`, thus not handling the error at the socket level, but the WebRTC connection level, thus only closing a WebRTC connection and not the whole listening socket.

> I am a bit thrown off by the fact that we get a _connection_ reset error on UDP socket, i.e. a connectionless protocol. Also, how should the receiving side detect a connection reset error if the other side (browser) just disappears (i.e. is closed) without sending any further data.
> 
> Off the top of my head I would explain this in the following way: on a UDP socket, send and receive share the same error slot. A previous send on the UDP socket caused an error, reported as a `ConnectionReset`. Our call to `poll_recv_from` surfaces that error from the previous send.
> 
> Any alternative explanations?

We might be sending a UDP datagram to a remote (e.g. the just closed browser tab). The remote is no longer listening on the UDP port we specified. The IP stack of the remote sends back an ICMP _Port Unreachable_. Our local `UdpSocket` surfaces the unreachable port error as a `ConnectionReset` in the call to `poll_recv_from`.

@DougAnderson444 do you see the same error (i.e. `ConnectionReset`) when you call `self.udp_sock.take_error()` right before the call to `self.udp_sock.poll_recv_from`?

**Comment by DougAnderson444** [comment]
> First off, great work @DougAnderson444 for debugging and proposing a patch. This is great help. I am sorry you ran into this. Thank you!

Happy to help, I'd really like this design to work!

> (As an aside, also great to see more usage of the WebRTC implementation in rust-libp2p.)

@mxinden I'm bullish on Rust 🦀 and WebRTC is a potentially great peer to peer solution, so I'm happy that we even have this option to work with! And yes, I am looking forward to the wasm web-sys Transport for the Browser-to-Browser version as well 😃 

<hr />

Your insights are great so please keep thinking out loud, I just started using WebRTC transport this week and had not gone that far down the UDP rabbit hole. I was just seeing the error bubble up through the Swarm as an Operating System error:

<details>
    <summary>UDPMux(Os { code: 10054 ...</summary>
    error: UDPMux(Os { code: 10054, kind: ConnectionReset, message: "An existing connection was forcibly closed by the remote host." }) })
</details>

I'm on Windows and [Code 10054 is a windows sockets error code 🔗 ](https://learn.microsoft.com/en-us/windows/win32/winsock/windows-sockets-error-codes-2#:%7E:text=WSAECONNRESET,10054)

So if we _do_ have more control over it, that would be great.

I inserted this right before `self.udp_sock.poll_recv_from`:

```rs
if let Ok(Some(err)) = self.udp_sock.take_error() {
    log::debug!("udp_sock take_error {err:?}");
}

match self.udp_sock.poll_recv_from(cx, &mut read) {
...
```

but it does not get called:

```cli
[2023-03-09T19:21:00Z DEBUG libp2p_webrtc::tokio::udp_mux] ConnectionReset by remote client Os { code: 10054, kind: ConnectionReset, message: "An existing connection was forcibly closed by the remote host." }
[2023-03-09T19:21:03Z INFO  libp2p_webrtc::tokio::substream::drop_listener] Substream 1 dropped without graceful 
close, sending Reset
[2023-03-09T19:21:03Z DEBUG libp2p_webrtc::tokio::connection] a DropListener failed: Stream closed
❌⛓️ Connection Closed to: 12D3KooWLLrBgqc5f2BVQ3G5ws39oxXUCeGyJfwQGrmYb1qMdKyX
```

File: transports/webrtc/CHANGELOG.md
Comment (mxinden) [inline transports/webrtc/CHANGELOG.md]:
```suggestion
- Gracefully handle `ConnectionReset` error on individual connections, avoiding shutdown of the entire listener upon disconnect of a single client.
```

This patch does not handle all errors gracefully.

File: transports/webrtc/CHANGELOG.md
Comment (mxinden) [inline transports/webrtc/CHANGELOG.md:1]: Can you bump the `transports/webrtc/Cargo.toml` version?

@thomaseizinger should we release a patch release of `libp2p` with a precise `libp2p-webrtc = { version = "=0.4.0-alpha.3", ...` as well, or just require users to import `libp2p-webrtc` `=0.4.0-alpha.3` individually?

File: transports/webrtc/src/tokio/udp_mux.rs
Comment (mxinden) [inline transports/webrtc/src/tokio/udp_mux.rs:404]:
```suggestion
                            log::debug!("ConnectionReset on socket: {err:?}")
```

I am not sure we can actually assume that the remote itself reset the connection. It might as well be in the networking stack of our machine or any router in between. Please correct me if I am wrong.

**Review by mxinden — COMMENTED** [review]
I am in favor of moving forward with the current solution, unless others have alternative suggestions (//CC @thomaseizinger and @melekes).

**Comment by mxinden** [comment]
> > (As an aside, also great to see more usage of the WebRTC implementation in rust-libp2p.)
> 
> @mxinden I'm bullish on Rust crab and WebRTC is a potentially great peer to peer solution, so I'm happy that we even have this option to work with! And yes, I am looking forward to the wasm web-sys Transport for the Browser-to-Browser version as well smiley

Yes, looking forward to browser-to-browser rust-libp2p via WASM as well. Let me know in case you would like to work on that.

Also, in case you want to get more involved, we do an open maintainers call every other week, see https://github.com/libp2p/rust-libp2p/discussions/.

**Comment by mxinden** [comment]
> I didn't write a test for this -- How do you write a test that fires a `std::io::error` for `ErrorKind::ConnectionReset` on the `udp_sock`?

I don't know how, other than making `UDPMuxNewAddr` generic over the `UdpSocket` and injecting a fake. I am not sure it is worth it. I am fine accepting this pull request without a test. Objections by others?

File: transports/webrtc/CHANGELOG.md
Comment (DougAnderson444) [reply transports/webrtc/CHANGELOG.md:1]: Bump complete.

File: transports/webrtc/src/tokio/udp_mux.rs
Comment (DougAnderson444) [reply transports/webrtc/src/tokio/udp_mux.rs:404]: I was originally just following the error message wording, which said: "remote host"

But I guess the prelude wording is kind of redundant as we're outputting the whole error message anyway

File: transports/webrtc/CHANGELOG.md
Comment (DougAnderson444) [reply transports/webrtc/CHANGELOG.md:1]: CI is breaking without `libp2p/Cargo.toml` being bumped to `0.4.0-alpha.3` too (so I bumped it to eliminate the red ❌s)

**Review by melekes — APPROVED** [review]
👍

**Comment by thomaseizinger** [comment]
> I am in favor of moving forward with the current solution, unless others have alternative suggestions (//CC @thomaseizinger and @melekes).

No objections.

**Comment by DougAnderson444** [comment]
Someone with write access (who isn't me) will have to resolve the conflicts

**Review by DougAnderson444 — COMMENTED** [review]
Trying to clear everything to make Github happy

**Comment by thomaseizinger** [comment]
> Someone with write access (who isn't me) will have to resolve the conflicts 

Why can you not resolve the conflicts on your own branch?

**Comment by mxinden** [comment]
I tried to resolve the merge conflicts, but seems like GitHub does not let me push to your forks `master` (see source of PR) branch.

**Review by mxinden — APPROVED** [review]

**Comment by mxinden** [comment]
> I tried to resolve the merge conflicts, but seems like GitHub does not let me push to your forks `master` (see source of PR) branch.

Worked from my local machine.

File: transports/webrtc/src/tokio/udp_mux.rs
Comment (DougAnderson444) [reply transports/webrtc/src/tokio/udp_mux.rs:401]: I made these flatten changes, marked this as resolved, yet Github still shows as outstanding. I'm a bit confused as to why...

**Comment by mxinden** [comment]
Failure on CI step ` Continuous integration / Ensure that `Cargo.lock` is up-to-date (pull_request)` is a false positive, see https://github.com/libp2p/rust-libp2p/pull/3587#discussion_r1132652573. Mergify should still be able to merge, given that the CI step is not yet marked as required.

**Comment by DougAnderson444** [comment]
I see there is "1 change requested" (the one from @thomaseizinger about the flat pattern matching) that is blocking mergify? Even though I did make it flat, and marked as resolved, so I'm a bit confused as to why it's still blocking the merge.

File: transports/webrtc/src/tokio/udp_mux.rs
Comment (mxinden) [reply transports/webrtc/src/tokio/udp_mux.rs:401]: As far as I can tell, this is due to the line above, namely the original line, not having changed. I think we are all good here. Thanks @DougAnderson444 for circling back on it.

**Comment by mxinden** [comment]
@mergify refresh

**Comment by mxinden** [comment]
Thanks @DougAnderson444 for the work here!

**Pull request merged**
